### ISSUE TITLE
Enum support

### ISSUE THREAD

**Issue by n-doton** [realm/realm-dart#681]
Is enum support planned? As far as I can tell, its currently not a supported type.
https://pub.dev/documentation/realm_common/latest/realm_common/RealmPropertyType.html

Thanks in advance!

**Comment by lukaspili** [comment]
See #579. Enhanced enum should make this trivial to support (from/to a string value internally)

**Comment by desistefanova** [comment]
Right, enums are not currently supported, but the value is always stored as int (or whatever is the type of the enum value) in the realm. But, yes for sure, we can consider supporting enums at least in Dart RealmModel and RealmModel generator.

**Comment by dotjon0** [comment]
We use enums all over our Flutter applications as it really helps us deliver fault tolerant applications and a great developer readability/experience. as MongoDB Atlas supports enums out of the box, it also makes a tone of sense to support enums in Flutter/Dart also. Will look forward to updates on this.

**Comment by dotjon0** [comment]
@desistefanova some further thoughts;

In MongoDB it's common for enum values to be in Snake Case (All Caps) i.e. **USER_LOGIN_COUNT**. In Flutter enums are in camelCase i.e. **userLoginCount**. So it may be useful to have some automatic case conversion for enum 'values' as a configurable option in Flutter Realm - if you see our other feature request here https://github.com/realm/realm-dart/issues/703 regarding https://pub.dev/packages/json_serializable#build-configuration which touches on this (different use case) where you can pass in configurable build rules via build.yaml

Many thanks again!

**Comment by hlazarpesic** [comment]
Any update here? Can we use enums in RealmModels?

**Comment by dotjon0** [comment]
Just checking in to see if enums are planned?

**Comment by hlazarpesic** [comment]
@desistefanova ☝️🙂

**Comment by desistefanova** [comment]
Supporting enums is in our list, but I can only say that it is not planned to be released soon. There are many others tasks before it.

**Comment by hlazarpesic** [comment]
Thanks for answer @desistefanova

**Comment by dotjon0** [comment]
Any updates for when this may land?

**Comment by nirinchev** [comment]
No - this requires core support which is tracked in this issue: https://github.com/realm/realm-core/issues/6159

**Comment by nielsenko** [comment]
@dotjon0 The above looks reasonable, but you will have to pass a `String` to the ctor.
```dart
import 'package:realm/realm.dart';

part 'main.realm.dart';

@RealmModel()
class _Car {
  @MapTo("type")
  late String _type;
  CustomType get type => CustomType.values.byName(_type);
  set type(final CustomType value) => _type = value.name;
}

enum CustomType { enabled, disabled }

final realm = Realm(Configuration.local([Car.schema]));

void main() {
  realm.write(() {
    // best bet right now ..
    realm.add(Car(CustomType.enabled.name));
    // or
    realm.add(Car('')..type = CustomType.enabled);
  });
}
```

But I realize there is bug when combining private fields, with either optional values, or named arguments.  
The following fx will generate invalid Dart code:
```dart
@RealmModel()
class _Car {
  @MapTo("type")
  late int _type = 0; // <-- will surface a bug :-/
  CustomType get type => CustomType.values[_type];
  set type(final CustomType value) => _type = value.index;
}
```

**Comment by nielsenko** [comment]
@dotjon0 The realm generator doesn't know the relation between the field `_type` and the property getter/setter pair `type`. While there may exist a convention for how such a relation-ship should look, there is no guarantee they are actually related at all.

What the generator does is just realizing that `type` is not a field, so it disregards looking further at it, but that `_type` is and (since ^2.1.0) allows that _private_ field to be stored. 

Hence you need to use the type of `_type` (`String`) in the ctor call and cannot use the enum type of `type` (`CustomType`). Sorry for the confusion the overloaded use of the word _type_ may cause, but it wasn't I who named these 😉.

You can improve the DX a bit, by using default values once we release #1663. Then you will be able to write:
```dart
@RealmModel()
class _Car {
  @MapTo("type")
  late String _type = CustomType.enabled.name; // or whatever default you prefer
  CustomType get type => CustomType.values.byName(_type);
  set type(final CustomType value) => _type = value.name;
}
// ...
void main() {
  realm.write(() {
    realm.add(Car()); // car with type field set to default value, ie. enabled
    realm.add(Car()..type = CustomType.enabled); // explicitly set type
  });
}
```
or
```dart
@RealmModel()
class _Car {
  @MapTo("type")
  late int _type = 0;
  CustomType get type => CustomType.values[_type];
  set type(final CustomType value) => _type = value.index;
}
```
The first is safer, as it is a lot easier to inadvertently change the _index_ value of an enum, than its _name_.

**Comment by nielsenko** [comment]
@dotjon0 Not currently. We need some way of registering conversion functions so that you can use the custom type at the field level, and have the generator write the code that invokes the conversions.

**Comment by nirinchev** [comment]
The problem with an SDK-only approach is that it breaks some fundamental assumptions that Realm is built on - like a property getter will never throw. Since Realm and Atlas don't enforce the valid enum values, there's nothing preventing someone from adding an Atlas document with `type: "crash-me"`, which will then result in exceptions being thrown when the property is read on the client. This is why we're overly cautious about adding what appears to be first-party support for enums - for someone not familiar with the relationship between Realm Core, Device Sync, and Atlas, it may appear like they can safely use the feature, when in reality, someone can break all their users by inserting an invalid enum value.

While we understand it's some extra work and somewhat annoying to have to manually do stuff like that, we have historically taken a conservative approach when designing our public API to have a good mix of usability and safety. While I don't expect us to add built-in support for enums before we're ready to commit to it across the entire stack, we'll be looking to improve the developer experience of customizing the models for dart - e.g. by allowing developers to ignore fields from the generated constructors or by allowing them to supply their own constructors. We don't have a timeframe for when such improvements could land, but we appreciate the feedback and take it into consideration during our planning meetings.

**Comment by nirinchev** [comment]
MongoDB Atlas does support enum values, but that is not well supported across the other products involved here and unfortunately, we can't make the assumption that users have defined a json schema on Atlas.

Regarding setters - they do work, the problem is that right now the generated constructor doesn't use the custom property, but rather the private one, which has the incorrect type (i.e. string rather than `CustomType`). This could be alleviated by what I mentioned above - either allowing you to opt out of certain fields in the generated ctor and opt-in to others or by allowing you to specify a custom constructor:

```dart
@RealmModel()
class _Car {
  @MapTo("type")
  @ConstructorOmit()
  late String _type;

  CustomType get type => CustomType.values.byName(_type);

  @ConstructorInclude()
  set type (final CustomType value) => _type = value.name;
}

// this generates
class Car extends _Car {
  Car({ CustomType type }) {
    this.type = type;
  }
}
```

Or alternatively, allow you somehow to specify the ctor (don't have a proposal for a good syntax here). Again, I'm mostly spitballing here and may be overlooking complexities, which will be addressed when we actually get to working on that, but right now we have higher priorities that we need to tackle.

**Comment by nirinchev** [comment]
That is correct - those are my (very preliminary) thoughts on how we could approach this. There's certainly a ton of minor details I'm overlooking and we'll probably want to figure out nicer names than those, but I don't see why it shouldn't work. Though, in the interest of transparency, I'd like to reiterate that we have several higher priority/higher impact tasks we need to tackle before we can look into generator customizations (notably, several web-related tickets, one of which I believe is impacting you as well).
